v2/storage/deferred/deferredcarwriter.go

```diff
@@ -0,0 +1,177 @@
+	roots     []cid.Cid
+	outPath   string
+	outStream io.Writer
+
+	lk    sync.Mutex
+	f     *os.File
+	w     carstorage.WritableCar
+	putCb []putCb
+	opts  []carv2.Option
+}
+
+// NewDeferredCarWriterForPath creates a DeferredCarWriter that will write to a
+// file designated by the supplied path. The file will only be created on the
+// first Put() operation.
+//
+// No options are supplied to carstorage.NewWritable by default, add
+// the car.WriteAsCarV1(true) option to write a CARv1 file.
+func NewDeferredCarWriterForPath(outPath string, roots []cid.Cid, opts ...carv2.Option) *DeferredCarWriter {
+	return &DeferredCarWriter{roots: roots, outPath: outPath, opts: opts}
+}
+
+// NewDeferredCarWriterForStream creates a DeferredCarWriter that will write to
+// the supplied stream. The stream will only be written to on the first Put()
+// operation.
+//
+// The car.WriteAsCarV1(true) option will be supplied by default to
+// carstorage.NewWritable as CARv2 is not a valid streaming format due to the
+// header.
+func NewDeferredCarWriterForStream(outStream io.Writer, roots []cid.Cid, opts ...carv2.Option) *DeferredCarWriter {
+	opts = append([]carv2.Option{carv2.WriteAsCarV1(true)}, opts...)
+	return &DeferredCarWriter{roots: roots, outStream: outStream, opts: opts}
+}
+
+// OnPut will call a callback when each Put() operation is started. The argument
+// to the callback is the number of bytes being written. If once is true, the
+// callback will be removed after the first call.
+func (dcw *DeferredCarWriter) OnPut(cb func(int), once bool) {
+	if dcw.putCb == nil {
+		dcw.putCb = make([]putCb, 0)
+	}
+	dcw.putCb = append(dcw.putCb, putCb{cb: cb, once: once})
+}
+
+// Has returns false if the key was not already written to the CAR output.
+func (dcw *DeferredCarWriter) Has(ctx context.Context, key string) (bool, error) {
+	dcw.lk.Lock()
+	defer dcw.lk.Unlock()
+
+	if dcw.w == nil { // shortcut, haven't written anything, don't even initialise
+		return false, nil
+	}
+
+	writer, err := dcw.writer()
+	if err != nil {
+		return false, err
+	}
+
+	return writer.Has(ctx, key)
+}
+
+// Put writes the given content to the CAR output stream, creating it if it
+// doesn't exist yet.
+func (dcw *DeferredCarWriter) Put(ctx context.Context, key string, content []byte) error {
+	dcw.lk.Lock()
+	defer dcw.lk.Unlock()
+
+	if dcw.putCb != nil {
+		// call all callbacks, remove those that were only needed once
+		for i := 0; i < len(dcw.putCb); i++ {
+			cb := dcw.putCb[i]
+			cb.cb(len(content))
+			if cb.once {
+				dcw.putCb = append(dcw.putCb[:i], dcw.putCb[i+1:]...)
+				i--
+			}
+		}
+	}
+
+	// first Put() call, initialise writer, which will write a CAR header
+	writer, err := dcw.writer()
+	if err != nil {
+		return err
+	}
+
+	return writer.Put(ctx, key, content)
+}
+
+// writer()
+func (dcw *DeferredCarWriter) writer() (carstorage.WritableCar, error) {
+	if dcw.w == nil {
+		outStream := dcw.outStream
+		if outStream == nil {
```

Comment:
if i write after close, is there a potential for this code path to open a file?

Reply:
Yep, there is. There's `closed` handling in the parent writer, but there are cases where it matters here. Added a bool and some checks for it.